test/testdrive/expected_group_size_tuning.td

```diff
@@ -0,0 +1,280 @@
+        WHERE l2.l_partkey = l1.l_partkey
+          AND l2.l_suppkey = l1.l_suppkey
+        ORDER BY l2.l_extendedprice DESC
+        LIMIT 1)
+  FROM lineitem l1
+  GROUP BY l1.l_partkey, l_suppkey;
+
+> CREATE MATERIALIZED VIEW lineitem_by_partkey AS
+  SELECT l1.l_partkey,
+        MAX(l1.l_extendedprice),
+        (SELECT l2.l_quantity FROM lineitem l2
+        WHERE l2.l_partkey = l1.l_partkey
+        ORDER BY l2.l_extendedprice DESC
+        LIMIT 1)
+  FROM lineitem l1
+  GROUP BY l1.l_partkey;
+
+> CREATE MATERIALIZED VIEW lineitem_by_suppkey AS
+  SELECT l1.l_suppkey,
+        MAX(l1.l_extendedprice),
+        (SELECT l2.l_quantity FROM lineitem l2
+        WHERE l2.l_suppkey = l1.l_suppkey
+        ORDER BY l2.l_extendedprice DESC
+        LIMIT 1)
+  FROM lineitem l1
+  GROUP BY l1.l_suppkey;
+
+# Check that we know the values of the perfect hints for TPC-H.
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 AS perfect_hint
+  FROM (
+        SELECT l_orderkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_orderkey
+  );
+15
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_partkey, l_suppkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_partkey, l_suppkey
+  );
+255
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_partkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_partkey
+  );
+255
+
+> SELECT pow(16, ceil(log(16, MAX(group_size)))) - 1 perfect_hint
+  FROM (
+        SELECT l_suppkey, COUNT(*) AS group_size
+        FROM lineitem
+        GROUP BY l_suppkey
+  );
+4095
+
+> WITH operators AS (
+  SELECT
+      dod.dataflow_id,
+      dor.id AS region_id,
+      dod.id,
+      ars.records
+  FROM
+      mz_internal.mz_dataflow_operator_dataflows dod
+      JOIN mz_internal.mz_dataflow_addresses doa
+          ON dod.id = doa.id
+      JOIN mz_internal.mz_dataflow_addresses dra
+          ON CAST(dra.address AS text) = CAST(doa.address[:list_length(doa.address) - 1] as text)
+      JOIN mz_internal.mz_dataflow_operators dor
+          ON dor.id = dra.id
+      JOIN mz_internal.mz_arrangement_sizes ars
+          ON ars.operator_id = dod.id
+  WHERE
+      dod.name = 'Arranged TopK input'
+      OR dod.name = 'Arranged MinsMaxesHierarchical input'
+      OR dod.name = 'Arrange ReduceMinsMaxes'
+  ),
+  levels AS (
+      SELECT o.dataflow_id, o.region_id, COUNT(*) AS levels
+      FROM operators o
+      GROUP BY o.dataflow_id, o.region_id
+  ),
+  pivot AS (
+      SELECT
+          o1.dataflow_id,
+          o1.region_id,
+          o1.id,
+          o1.records
+      FROM operators o1
+      WHERE
+          o1.id = (
+              SELECT MIN(o2.id)
+              FROM operators o2
+              WHERE
+                  o2.dataflow_id = o1.dataflow_id
+                  AND o2.region_id = o1.region_id
+              OPTIONS (EXPECTED GROUP SIZE = 8)
```

Comment:
How important is this optimizer hint for query performance?

And how is the 8 chosen? Are there at most 8 levels in a reduce/topk?

Reply:
The reason for 8 here is that this is the number of levels we'd expect to get in a region with a min/max or top-k arrangement hierarchy in the absence of a query hint:  https://github.com/MaterializeInc/materialize/blob/6daf1aa3cd67ee3cb2cf5a4674887c2ff09a77b6/src/compute-client/src/plan/mod.rs#L2236-L2254

My feeling was that since we were trying to tune this hint with this query, then the query ought to be tuned as well. :-) I did not observe much of difference in query response time with the TPC-H example provided in the test. That being said, the absolute response times are not large for the query at this point, since the input data is not very large.